src/storage/txn/scheduler.rs

```diff
@@ -610,38 +629,82 @@ impl<E: Engine, L: LockManager> Scheduler<E, L> {
                 pr,
                 lock_info,
                 lock_guards,
+                response_policy,
             }) => {
                 SCHED_STAGE_COUNTER_VEC.get(tag).write.inc();
 
                 if let Some((lock, is_first_lock, wait_timeout)) = lock_info {
                     scheduler.on_wait_for_lock(cid, ts, pr, lock, is_first_lock, wait_timeout);
                 } else if to_be_write.modifies.is_empty() {
-                    scheduler.on_write_finished(cid, pr, Ok(()), lock_guards, false, tag);
+                    scheduler.on_write_finished(
+                        cid,
+                        Some(pr),
+                        Ok(()),
+                        lock_guards,
+                        false,
+                        false,
+                        tag,
+                    );
                 } else {
-                    let proposed_cb: Option<ExtCallback> = if pipelined {
-                        // The normal write process is respond to clients and release latches
-                        // after async write finished. If pipelined pessimistic locking is enabled,
-                        // the process becomes parallel and there are two msgs for one command:
-                        //   1. Msg::PipelinedWrite: respond to clients
-                        //   2. Msg::WriteFinished: deque context and release latches
-                        // The order between these two msgs is uncertain due to thread scheduling
-                        // so we clone the result for each msg.
-                        let pipelined_write_pr = pr.maybe_clone().unwrap();
-                        let sched = scheduler.clone();
-                        let sched_pool = scheduler.get_sched_pool(priority).pool.clone();
-                        Some(Box::new(move || {
-                            sched_pool
-                                .spawn(async move {
-                                    fail_point!("scheduler_pipelined_write_finish");
-                                    // The write task is proposed to the raftstore successfully.
-                                    // Respond to client early.
-                                    sched.on_pipelined_write(cid, pipelined_write_pr, tag);
-                                })
-                                .unwrap()
-                        }))
-                    } else {
-                        None
-                    };
+                    let mut pr = Some(pr);
+                    let mut is_async_apply_prewrite = false;
+
+                    let (proposed_cb, committed_cb): (Option<ExtCallback>, Option<ExtCallback>) =
+                        match response_policy {
+                            ResponsePolicy::OnApplied => (None, None),
+                            ResponsePolicy::OnCommitted => {
+                                self.inner.store_pr(cid, pr.take().unwrap());
```

Comment:
But the committed/proposed callback is not guaranteed to be invoked (see [here](https://github.com/tikv/tikv/blob/d024af5618def4f6ec1fa94f872d880e952c243f/components/raftstore/src/store/peer.rs#L2072-L2078) and [here](https://github.com/tikv/tikv/blob/d024af5618def4f6ec1fa94f872d880e952c243f/components/raftstore/src/store/peer.rs#L1727-L1733)), and in this case the pr will be still processed in `on_write_finished`. But if we let the callback capture the pr, it will be hard to handle the case when the callback is not invoked.

Reply:
OK. Makes sense.